plugins/modules/source_control/gitlab/gitlab_group_members.py

```diff
@@ -165,98 +201,187 @@ def main():
+                if user_access_level == gitlab_user['access_level']:
+                    changed_users.append("User, '%s', is already a member in the group. No change to report" % gitlab_user['name'])
+                    changed_data.append({'gitlab_user': gitlab_user['name'], 'result': 'OK',
+                                         'msg': "User, '%s', is already a member in the group. No change to report" % gitlab_user['name']})
+                else:
+                    # update the access level for the user
+                    if not module.check_mode:
+                        group.update_user_access_level(members, gitlab_user_id, gitlab_user['access_level'])
+                    changed = True
+                    changed_users.append("Successfully updated the access level for the user, '%s'" % gitlab_user['name'])
+                    changed_data.append({'gitlab_user': gitlab_user['name'], 'result': 'CHANGED',
+                                         'msg': "Successfully updated the access level for the user, '%s'" % gitlab_user['name']})
+            else:
+                # remove the user from the group
+                if not module.check_mode:
+                    group.remove_user_from_group(gitlab_user_id, gitlab_group_id)
+                changed = True
+                changed_users.append("Successfully removed user, '%s', from the group" % gitlab_user['name'])
+                changed_data.append({'gitlab_user': gitlab_user['name'], 'result': 'CHANGED',
+                                     'msg': "Successfully removed user, '%s', from the group" % gitlab_user['name']})
+
+    # if state = present and purge_users set delete users which are in members having give access level but not in gitlab_users
+    if state == 'present' and purge_users:
+        for member in members:
+            if member.access_level in purge_users and member.username.upper() not in [name['name'].upper() for name in gitlab_users_access]:
+                group.remove_user_from_group(member.id, gitlab_group_id)
+                changed = True
+                changed_users.append("Successfully removed user '%s', from group. Was not in given list" % member.username)
+                changed_data.append({'gitlab_user': member.username, 'result': 'CHANGED',
+                                     'msg': "Successfully removed user '%s', from group. Was not in given list" % member.username})
+
+    if len(gitlab_users_access) == 1 and error:
+        # if single user given and an error occurred return error for list errors will be per user
+        module.fail_json(msg=changed_users[0])
+    elif error:
+        # iterate results, if all failed set to failed
+        single_good = False
+        for cd in changed_data:
+            if cd.result != 'FAILED':
+                single_good = True
+        if not single_good:
+            module.fail_json(msg='All tasks for all users given failed', result_data=changed_data)
```

Comment:
IMO the module should fail if there was at least one error. After all an error happened, the user deserves to know.

Reply:
You are right, Done.